tests/generic-java-signatures/outer-ref-elimination.scala

```diff
@@ -0,0 +1,36 @@
+// This test checks that references to outer classes in inner classes are
+// eliminated in some cases when they are not used. This is done is the
+// ExplicitOuter phase. See issue #19569 for discussions.
+
+object helper {
+  def test(cls: Class[?]) = println(cls.getDeclaredConstructors.toList)
+}
+
+import helper.test
+
+object T1 { class C1; test(classOf[C1]) }
+object T2 { new AnyRef { class C2; test(classOf[C2]) } }
+object T3 { def t3(): Unit = { class C3; test(classOf[C3]) } }
+object T4 { def t4(): Unit = new AnyRef { class C4; test(classOf[C4]) } }
+
+class T5 { class C5; test(classOf[C5]) } // outer ref currently not elided
```

Comment:
I think "currently" isn't correct (and thus misleading).  Classes that aren't defined in some kind of block are externally instantiable.  So they are assumed to be public API, which means a change in the class's implementation can't break the API of that class's constructor.

Reply:
Yeah, it's assumed that `T1$` is instantiated only once, by val T1, as opposed to multiple T5 instances.

Reply:
Comments updated.